File: src/test/java/me/saro/kotlin/java/MainTest.java
package me.saro.kotlin.java;

import io.jsonwebtoken.SignatureAlgorithm;
import java.lang.SecurityException;
import me.saro.jwt.JwtKeyManager;
import me.saro.jwt.impl.DefaultJwtKeyManager;
import me.saro.jwt.io.JwtBuilder;
import me.saro.jwt.io.JwtReader;
import me.saro.jwt.model.ClaimName;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Base64;

@DisplayName("[Java] JwtReader")
public class MainTest {
    @Test
    @DisplayName("normal")
    public void t1() {
        String name = "안녕 hello !@#$";
        String encode = "안녕 hello !@#$";

        JwtKeyManager jwtKeyManager = DefaultJwtKeyManager.create(SignatureAlgorithm.RS256);
        JwtBuilder builder = jwtKeyManager.getJwtBuilder();
        builder.claim("name", name);
        builder.encryptClaim("encode", encode);
        builder.setIssuedAtNow();
        builder.setExpireMinutes(30);
        builder.claim(ClaimName.id, "id");
        builder.claim(ClaimName.subject, "sub");
        builder.claim(ClaimName.issuer, "iss");
        String jwt = builder.build();

        System.out.println("- JWT");
        System.out.println(jwt);
        System.out.println("- header / payload");
        Arrays.asList(jwt.split("\\.")).stream().limit(2)
                .map(e -> new String(Base64.getDecoder().decode(e.replace('-', '+').replace('_', '/'))))
                .forEach(System.out::println);

        JwtReader jwtReader = jwtKeyManager.parse(jwt);

        Assertions.assertEquals(jwtReader.claim("name").toString(), name);
        Assertions.assertEquals(jwtReader.decryptClaim("encode").toString(), encode);
    }

    @Test
    @DisplayName("signature pass and error")
    public void t2() {
        JwtKeyManager m1 = DefaultJwtKeyManager.create(SignatureAlgorithm.RS256);
        JwtKeyManager m2 = DefaultJwtKeyManager.create(SignatureAlgorithm.RS256);

        String jwt = m1.getJwtBuilder().claim("text", "hello").build();

        Assertions.assertDoesNotThrow(() -> {
            JwtReader reader = m1.parse(jwt);
            System.out.println(reader.claim("text"));
        });

        Assertions.assertThrows(SecurityException.class, () -> {
            JwtReader reader = m2.parse(jwt);
            System.out.println(reader.claim("text"));
        });
    }
}

File: src/test/kotlin/me/saro/jwt/kotlin/MainTest.kt
//package me.saro.jwt.kotlin
//
//import io.jsonwebtoken.SignatureAlgorithm
//import me.saro.jwt.old.JwtKeyManager
//import me.saro.jwt.old.impl.DefaultJwtKeyManager.Companion.create
//import org.junit.jupiter.api.Assertions
//import org.junit.jupiter.api.DisplayName
//import org.junit.jupiter.api.Test
//import java.util.*
//
//@DisplayName("[Kotlin] MainTest")
//class MainTest {
//    @Test
//    fun normal() {
//        val name = "안녕 hello !@#$"
//        val encode = "안녕 hello !@#$"
//        val jwtKeyManager: JwtKeyManager = create(SignatureAlgorithm.RS256)
//        val jwt = jwtKeyManager.getJwtBuilder()
//            .claim("name", name)
//            .encryptClaim("encode", encode)
//            .issuedAtNow()
//            .expireMinutes(30)
//            .id("id")
//            .subject("sub")
//            .issuer("iss")
//            .build()
//
//        println("jwt: $jwt")
//
//        val header = String(Base64.getDecoder().decode(jwt.split(".")[0]))
//        println("header: $header")
//
//        val payload = String(Base64.getDecoder().decode(jwt.split(".")[1]))
//        println("payload: $payload")
//
//        val jwtReader = jwtKeyManager.parse(jwt)
//        Assertions.assertEquals(name, jwtReader.claim("name"))
//        Assertions.assertEquals(encode, jwtReader.decryptClaim("encode"))
//    }
//
//    @Test
//    fun `signature pass and error`() {
//        val m1: JwtKeyManager = create(SignatureAlgorithm.RS256)
//        val m2: JwtKeyManager = create(SignatureAlgorithm.RS256)
//        val jwt = m1.getJwtBuilder().claim("text", "hello").build()
//        Assertions.assertDoesNotThrow {
//            val reader = m1.parse(jwt)
//            println(reader.claim("text"))
//        }
//        Assertions.assertThrows(SecurityException::class.java) {
//            val reader = m2.parse(jwt)
//            println(reader.claim("text"))
//        }
//    }
//
//
//    @Test
//    fun example() {
//        // algorithm: RS256, key rotation queue size: 3, key rotation minutes: 30
//        // key stored [rotation minutes (30) * queue size (3)] = 90 minutes
//        val jwtKeyManager: JwtKeyManager = create(SignatureAlgorithm.RS256, 3, 30)
//
//        val jwt = jwtKeyManager.getJwtBuilder()
//            .encryptClaim("jti", "1234")
//            .subject("sub")
//            .issuer("iss")
//            .issuedAtNow()
//            .expireMinutes(30)
//            .build()
//        println("jwt: $jwt")
//
//        val header = String(Base64.getDecoder().decode(jwt.split(".")[0]))
//        println("header: $header")
//
//        val payload = String(Base64.getDecoder().decode(jwt.split(".")[1]))
//        println("payload: $payload")
//
//        val jwtReader = jwtKeyManager.parse(jwt)
//
//        val result = mapOf(
//            "id" to jwtReader.decryptClaim("jti"),
//            "subject" to jwtReader.subject,
//            "issuer" to jwtReader.issuer,
//            "issuedAt" to jwtReader.issuedAt,
//            "expire" to jwtReader.expire
//        )
//        println("result: $result")
//    }
//}